source/common/http/alternate_protocols_cache_impl.cc

```diff
@@ -2,22 +2,84 @@
 
 #include "source/common/common/logger.h"
 
+#include "quiche/spdy/core/spdy_alt_svc_wire_format.h"
+
 namespace Envoy {
 namespace Http {
+namespace {
+std::string originToString(const AlternateProtocolsCache::Origin& origin) {
+  return absl::StrCat(origin.scheme_, "://", origin.hostname_, ":", origin.port_);
+}
+} // namespace
+
+std::string AlternateProtocolsCacheImpl::protocolsToStringForCache(
+    const std::vector<AlternateProtocol>& protocols, TimeSource& /*time_source*/) {
+  if (protocols.empty()) {
+    return std::string("clear");
+  }
+  std::string value;
+  for (auto& protocol : protocols) {
+    if (!value.empty()) {
+      value.push_back(',');
+    }
+    absl::StrAppend(&value, protocol.alpn_, "=\"", protocol.hostname_, ":", protocol.port_, "\"");
+
+    // Note this is _not_ actually the max age, but the absolute time at which
+    // this entry will expire. protocolsFromString will convert back to ma.
+    absl::StrAppend(
+        &value, "; ma=",
+        std::chrono::duration_cast<std::chrono::seconds>(protocol.expiration_.time_since_epoch())
+            .count());
+  }
+  return value;
+}
 
-AlternateProtocolsCacheImpl::AlternateProtocolsCacheImpl(TimeSource& time_source)
-    : time_source_(time_source) {}
+absl::optional<std::vector<AlternateProtocolsCache::AlternateProtocol>>
+AlternateProtocolsCacheImpl::protocolsFromString(absl::string_view alt_svc_string,
+                                                 TimeSource& time_source, bool from_cache) {
+  std::vector<AlternateProtocol> protocols;
+  spdy::SpdyAltSvcWireFormat::AlternativeServiceVector altsvc_vector;
+  if (!spdy::SpdyAltSvcWireFormat::ParseHeaderFieldValue(alt_svc_string, &altsvc_vector)) {
+    return {};
+  }
+  for (const auto& alt_svc : altsvc_vector) {
+    MonotonicTime expiration;
+    if (from_cache) {
+      auto expire_time_from_epoch = std::chrono::minutes(alt_svc.max_age);
```

Comment:
Hm, I notice that on line 31 we seem to write ma as seconds. But here it looks like we parse it as minutes. Is that a bug? (I don't think so 'cause I think you have test coverage for it? Does the test coverage report 100% for this method?)

Reply:
Oh nice catch.  I was unit testing the non-cache conversion and there were indeed issues in the cache versioning which would be caught by the not-yet-done integration test.  fixed and tested.